### PR TITLE
Challenge 2.0

### PR DESCRIPTION
**Pull request by DOMINION-JOHN1**
started level 2 challenge. You can at occasion check out my work

### CODE REVIEW

File: CHALLENGE 2.0/def,multiple_3_calc.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+def mul_calc(num):
+    num=float(input("enter value here; "))
```
Comment (greatertomi) [inline CHALLENGE 2.0/def,multiple_3_calc.py]: Don't accept from this function, only use the `num` parameter.

File: CHALLENGE 2.0/def,multiple_3_calc.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    arrayA=range(3,100,5)
```
Comment (greatertomi) [inline CHALLENGE 2.0/def,multiple_3_calc.py]: You can safely remove this line and do something like this `for i in range(3,100,5)`

File: CHALLENGE 2.0/def,trian_calc.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+def trian_calc (A,B,C):
+    A=float(input("insert a number here; "))
+    B=float(input("insert a number here; "))
+    C=float(input("insert a number here; "))
```
Comment (greatertomi) [inline CHALLENGE 2.0/def,trian_calc.py]: Only use the functions paramaters. Remove these

File: CHALLENGE 2.0/def,multiple_3_calc.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+    for i in arrayA:
+        print(i%num==0)
+
```
Comment (greatertomi) [inline CHALLENGE 2.0/def,multiple_3_calc.py]: Your function does not return anything. I am expecting it to return the answer. Use `return` instead of `print`

File: CHALLENGE 2.0/def,multiple_3_calc.py (hunk 2 above)
Comment (greatertomi) [inline CHALLENGE 2.0/def,multiple_3_calc.py]: Are you sure `range(3,100,5)` would solve any array of numbers? Because I don't why you are using fixed numbers here.

File: CHALLENGE 2.0/def,trian_calc.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    if  A+B > C and  A+C > B and B+C > A:
+        print(True)
+    else:
+        print(False)
```
Comment (greatertomi) [inline CHALLENGE 2.0/def,trian_calc.py]: This function does not return anything. I am expecting it to return the answer. So don't use `print`, return the answer

**Review by greatertomi — CHANGES_REQUESTED** [review]

**Comment by DOMINION-JOHN1** [comment]
apologies for the delay, i had to finish some modules to be able to solve these challenges.  Do check out what i have being working on. should be done in a day or two.

**Comment by DOMINION-JOHN1** [comment]
for the part B of this challenge I created a new python file for each modification of the app. example: multi_tutor 1 means the first questions of part b

File: CHALLENGE 2.0/def,trian_calc.py (hunk 3 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/def,trian_calc.py]: alright i already removed the input function

File: CHALLENGE 2.0/def,trian_calc.py (hunk 5 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/def,trian_calc.py]: done!

File: CHALLENGE 2.0/def,multiple_3_calc.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+def mul_calc(num):
+    num=float(input("enter value here; "))
+    arrayA=range(3,100,5)
+    for i in arrayA:
+        print(i%num==0)
```
Comment (DOMINION-JOHN1) [inline CHALLENGE 2.0/def,multiple_3_calc.py]: oh this has being corrected

File: CHALLENGE 2.0/def,multiple_3_calc.py (hunk 2 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/def,multiple_3_calc.py]: yes , i have changed *arg  in the def function so it can collect  an any array of numbers

File: CHALLENGE 2.0/def,multiple_3_calc.py (hunk 1 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/def,multiple_3_calc.py]: resolved

File: CHALLENGE 2.0/def,multiple_3_calc.py (hunk 2 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/def,multiple_3_calc.py]: resolved.

File: CHALLENGE 2.0/def,multiple_3_calc.py (hunk 6 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/def,multiple_3_calc.py]: there was suposed to be an If statement inbetween

File: CHALLENGE 2.0/def,multiple_3_calc.py (hunk 4 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/def,multiple_3_calc.py]: Now resolved

File: CHALLENGE 2.0/def,trian_calc.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+def trian_calc ():
+    A=float(input("insert a number here; "))
+    B=float(input("insert a number here; "))
+    C=float(input("insert a number here; "))
```
Comment (greatertomi) [inline CHALLENGE 2.0/def,trian_calc.py]: You've not removed the input function. Do something like this and use the input function outside this triangle function
```suggestion
def triangle_calculation(side1, side2, side3):
   if  side1+side2 > side3 and  side1+side3 > side2 and side2+side3 > side1
   ...
   ```

File: CHALLENGE 2.0/def,trian_calc.py (hunk 5 above)
Comment (greatertomi) [reply CHALLENGE 2.0/def,trian_calc.py]: You have not resolved this. Use this kind of return

```suggestion
        return False
```

File: CHALLENGE 2.0/distance_calculator.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+def calculate_distance():
+    x1=float(input("enter value here; "))
+    y1=float(input("enter value here; "))
+    x2 = float(input("enter value here; "))
+    y2=float(input("enter value here; "))
```
Comment (greatertomi) [inline CHALLENGE 2.0/distance_calculator.py]: Don't use input inside this function. Do something like 

```suggestion
def calculate_distance(x1, y1, x2, y2, d):
    ...
```

File: CHALLENGE 2.0/def,trian_calc.py (hunk 3 above)
Comment (greatertomi) [reply CHALLENGE 2.0/def,trian_calc.py]: Not resolved

**Review by greatertomi — CHANGES_REQUESTED** [review]

**Comment by DOMINION-JOHN1** [comment]
Changes have being pushed to branch challenge _2.0

File: CHALLENGE 2.0/distance_calculator.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+def calculate_distance(x1,y1,x2,y2):
+    distance=(((x2-x1)**2) +((y2-y1)**2))**0.5
+    print(distance)
```
Comment (greatertomi) [inline CHALLENGE 2.0/distance_calculator.py]: Use `return` instead of `print`

File: CHALLENGE 2.0/distance_calculator.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+## call the function and input the desired variables
```
Comment (greatertomi) [inline CHALLENGE 2.0/distance_calculator.py]: Comment should be one "#"

File: CHALLENGE 2.0/distance_calculator.py [hunk 11]
```diff
@@ -0,0 +1,5 @@
...
+calculate_distance(4,5,6,7)
```
Comment (greatertomi) [inline CHALLENGE 2.0/distance_calculator.py]: You have to capture edge cases, don't just test the happy path scenario. Make sure these returns a proper answer

```
calculate_distance(4,5) -> throw an error that some values are missing
calculate_distance(4,5, 'b', '2') -> throw an error that you only accept integers
calculate_distance(4,5, '6', '2.56') -> convert the strings to number and calculate with them
```

File: CHALLENGE 2.0/find_multiples_in_array.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+def get_multiples(arrayA,num):
+    arrayA = [2, 4, 6, 8, 10, 12, 14]
+    num = 2
```
Comment (greatertomi) [inline CHALLENGE 2.0/find_multiples_in_array.py]: Pass the `arrayA` and `num` from outside the function, because here you are not using the parameters.

File: CHALLENGE 2.0/find_multiples_in_array.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+    return [x for x in arrayA if x % num == 0]
```
Comment (greatertomi) [inline CHALLENGE 2.0/find_multiples_in_array.py:6]: I like this list comprehension. Smart stuff

File: CHALLENGE 2.0/find_multiples_in_array.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+result = get_multiples('arrayA','num')
```
Comment (greatertomi) [inline CHALLENGE 2.0/find_multiples_in_array.py]: Passing strings that you didn't use into the function doesn't make sense.

File: CHALLENGE 2.0/unique_string.py [hunk 15]
```diff
@@ -0,0 +1,4 @@
+def remove_duplicates(*args):
+    print(set(args))
```
Comment (greatertomi) [inline CHALLENGE 2.0/unique_string.py]: use `return` instead of `print`

File: CHALLENGE 2.0/unique_string.py [hunk 16]
```diff
@@ -0,0 +1,4 @@
+def remove_duplicates(*args):
```
Comment (greatertomi) [inline CHALLENGE 2.0/unique_string.py]: This function should recieve a string, not *args. This function can't pass the testcases in the question

```
remove_duplicates('AABBBCCCDEEEE')
remove_duplicates('GHIIIKKL')
```

File: CHALLENGE 2.0/numbers_that_form_triangle.py [hunk 17]
```diff
@@ -0,0 +1,11 @@
+def triangle_num(a, b, c):
```
Comment (greatertomi) [inline CHALLENGE 2.0/numbers_that_form_triangle.py]: You always assume your function would receive a happy-path parameter. Could you validate the inputs here before doing the calculation?

**Review by greatertomi — CHANGES_REQUESTED** [review]

File: CHALLENGE 2.0/numbers_that_form_triangle.py [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+    if (a + b > c) and (b + c > a) and (c + a > b):
```
Comment (greatertomi) [inline CHALLENGE 2.0/numbers_that_form_triangle.py]: It would also be good to improve on the variable naming instead of a, b and c

File: CHALLENGE 2.0/distance_calculator.py (hunk 11 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/distance_calculator.py]: done with the third
i am not too clear on this
for the first two you want me to add them as an error message?

File: CHALLENGE 2.0/find_multiples_in_array.py (hunk 13 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/find_multiples_in_array.py:6]: Finally! my head is swelling. Thanks

File: CHALLENGE 2.0/find_multiples_in_array.py (hunk 14 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/find_multiples_in_array.py]: done

File: CHALLENGE 2.0/unique_string.py (hunk 15 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/unique_string.py]: true! corrected .

File: CHALLENGE 2.0/numbers_that_form_triangle.py (hunk 17 above)
Comment (DOMINION-JOHN1) [reply CHALLENGE 2.0/numbers_that_form_triangle.py]: i dont grab this. By validate do you mean i  should make the function accept other data type?

File: CHALLENGE 2.0/unique_string.py [hunk 19]
```diff
@@ -0,0 +1,5 @@
+def remove_duplicates(string_list):
+    return set(string_list)
+
+
+print(remove_duplicates('namedominionjohneleojo'))
```
Comment (greatertomi) [inline CHALLENGE 2.0/unique_string.py:5]: This returns `{'m', 'h', 'a', 'i', 'l', 'o', 'j', 'e', 'd', 'n'}`. It should return a string like "adelhijmno"

File: CHALLENGE 2.0/guess_the_number_calc.py
Comment (greatertomi) [inline CHALLENGE 2.0/guess_the_number_calc.py:1]: Why do I always get the answer after every second try? The generated number was 368, for example. I entered 500 first try, then 250 second try and the app said I was correct.

Check the logic of this

File: CHALLENGE 2.0/maths_multiplication_tutor_legend.py
Comment (greatertomi) [inline CHALLENGE 2.0/maths_multiplication_tutor_legend.py:1]: - This has all the above errors I wrote about level2 and level3. 
- In division, the app does not mark my right answer and maybe you should approximate division answers to 2 decimal places (don't forget to notify the users). 
- Selecting mixture questions only brings up division problems

File: CHALLENGE 2.0/maths_multiplication_tutor_level2.py
Comment (greatertomi) [inline CHALLENGE 2.0/maths_multiplication_tutor_level2.py:1]: Correct answer always throws "Excellent!" and the wrong answer always throws “No. Keep trying“. The responses are not varied.

File: CHALLENGE 2.0/multiplication_tutor_level3.py
Comment (greatertomi) [inline CHALLENGE 2.0/multiplication_tutor_level3.py:1]: - ~Why does it allow me to reanswer questions I failed?~
- It doesn’t grade. I can answer more than 10 questions. It should restrict to only 10 questions

I feel you don’t even check the requirements of the question. After coding, check if your code meets the requirements.

**Review by greatertomi — CHANGES_REQUESTED** [review]

**Review by greatertomi — APPROVED** [review]

**Comment by DOMINION-JOHN1** [comment]
Thanks sir for the approval
